src/modules/prometheus.ts

```diff
@@ -0,0 +1,104 @@
+import http from "http"
+
+import Prometheus from "prom-client"
+import promBundle from "express-prom-bundle"
+
+import { searchTypes } from "../../src/constants/media"
+
+import type { ServerMiddleware, Module } from "@nuxt/types"
+
+let metricsServer: null | http.Server = null
+
+const bypassMetricsPathParts = [
+  /**
+   * Exclude static paths. Remove once we've moved static file
+   * hosting out of the SSR server's responsibilities.
+   */
+  "_nuxt",
+  /**
+   * Only include these paths in development. They do not exist in production
+   * so we can happily skip the extra iterations these path parts introduce.
+   */
+  ...(process.env.NODE_ENV === "development" ? ["__webpack", "sse"] : []),
+]
+
+const bundle = promBundle({
+  // We serve the Prometheus metrics on a separate port for production
+  // and if we let express-prom-bundle register the `/metrics` route then
+  // it exposes all the metrics on the publicly accessible port
+  autoregister: false,
+  promClient: {
+    // Set this to an empty object to pass falsy check in express-prom-bundle
+    // and default metrics get enabled with the default prom-client configuration
+    collectDefaultMetrics: {},
+  },
+  buckets: [0.03, 0.3, 0.5, 1, 1.5, 2, 5, 10, Infinity],
+  bypass: (req) =>
+    bypassMetricsPathParts.some((p) => req.originalUrl.includes(p)),
+  includeMethod: true,
+  includePath: true,
+  normalizePath: [
+    ...searchTypes.map(
+      // Normalize single result pages with IDs in the path
+      (t) => [`/${t}/.*`, `/${t}/#id`] as [string, string]
+    ),
+  ],
+  /**
+   * promBundle creates an Express middleware function, which is type-incompatible
+   * with Nuxt's "connect" middleware functions. I guess they _are_ compatible,
+   * in actual code, or maybe just in the particular implementation of the prometheus
+   * bundle middleware. In any case, this cast is necessary to appeas TypeScript
+   * without an ugly ts-ignore.
+   */
+}) as unknown as ServerMiddleware
+
+const PrometheusModule: Module = function () {
+  // Nuxt types don't fully type `this` parameter, this.nuxt is any.
```

Comment:
FWIW you can use [this feature (no pun intended)](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#specifying-the-type-of-this-for-functions) to type `this`.

Reply:
I saw the example of typing `this` in the Nuxt's Module type:
https://github.com/nuxt/nuxt/blob/853439ddf945b6ee358d5912e6e5a68009633c27/packages/types/config/module.d.ts#L29
But the `this.nuxt`'s type there is `any` :( . I am not sure it's worth the effort to add the type (and search for what exactly needs to be added) since the modules will need to be refactored for Nuxt 3.

Reply:
You're right, it must have been WebStorm complaining for me. I removed the `@ts-ignore` comments, and added type to `app` parameter from the `nuxt/types` repository hooks module (`ConnectServer`)